### ISSUE TITLE
pydantic.validate_call is converting strings to decimals when string is one of the allowed types

### ISSUE THREAD

**Issue by hb2638** [pydantic/pydantic#10142]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

pydantic.validate_call is converting strings to decimals when string is one of the allowed types

No conversion should be done when an exact type match is listed in the allowed types.

So logic should be, see if exact type already listed in union, if so, return the value unaltered (no conversion/coercion)... otherwise do the heuristics that you normally do.

Not sure if this was fixed in https://github.com/pydantic/pydantic-core/pull/1332 . If so, was that code deployed to pypi?

The example code outputs

```
<class 'str'>
<class 'decimal.Decimal'>
10031853
10031853
```

### Example Code

```Python
import decimal
import typing
import uuid
from datetime import date, datetime

import pydantic


@pydantic.validate_call(config=pydantic.ConfigDict(arbitrary_types_allowed=True))
def echo(value: date|datetime|float|decimal.Decimal|int|str|bool|None|typing.Callable|uuid.UUID) -> object:
    return value


expected = "10031853"
actual = echo(expected)
print(type(expected))
print(type(actual))
print(expected)
print(actual)
```

### Python, Pydantic & OS Version

```Text
>python -c "import pydantic.version; print(pydantic.version.version_info())"
             pydantic version: 2.8.2
        pydantic-core version: 2.20.1
          pydantic-core build: profile=release pgo=true
                 install path: D:\code\venv\Lib\site-packages\pydantic
               python version: 3.12.1 (tags/v3.12.1:2305ca5, Dec  7 2023, 22:03:25) [MSC v.1937 64 bit (AMD64)]
                     platform: Windows-11-10.0.22631-SP0
             related packages: typing_extensions-4.12.2
                       commit: unknown

I also force install the latest version of pydantic-core... bug still there

>python -c "import pydantic.version; print(pydantic.version.version_info())"
             pydantic version: 2.8.2
        pydantic-core version: 2.21.0
          pydantic-core build: profile=release pgo=true
                 install path: D:\code\venv\Lib\site-packages\pydantic
               python version: 3.12.1 (tags/v3.12.1:2305ca5, Dec  7 2023, 22:03:25) [MSC v.1937 64 bit (AMD64)]
                     platform: Windows-11-10.0.22631-SP0
             related packages: typing_extensions-4.12.2
                       commit: unknown
```

**Comment by hb2638** [comment]
It works when I make **str** the first thing in the union.

E.x.: def echo(value: str|date|datetime|float|decimal.Decimal|int|bool|None|typing.Callable|uuid.UUID) -> object:

**Comment by sydney-runkle** [comment]
Hi @hb2638,

If you use `strict=True`, you get the expected behavior:

```py
import decimal
import typing
import uuid
from datetime import date, datetime

import pydantic


@pydantic.validate_call(config=pydantic.ConfigDict(arbitrary_types_allowed=True, strict=True))
def echo(value: date|datetime|float|decimal.Decimal|int|str|bool|None|typing.Callable|uuid.UUID) -> object:
    return value


expected = "10031853"
actual = echo(expected)
print(type(expected))
print(type(actual))
print(expected)
print(actual)
```

I think the real question here is why do smart unions select `decimal.Decimal` over `str` for `str` type data in lax mode, which has been raised [here](https://github.com/pydantic/pydantic/issues/9795).

A more minimal reproducible example:

```py
import decimal

from pydantic import TypeAdapter


expected = '10031853'

ta = TypeAdapter(decimal.Decimal | str)
actual = ta.validate_python(expected)

print(type(expected))
print(type(actual))
print(expected)
print(actual)
```

Going to close this as a duplicate.

**Comment by hb2638** [comment]
thank you for looking into this!
I wish other projects were as on top of issues as this one.

**Comment by sydney-runkle** [comment]
Sure thing. I'm digging around for a fix now, hopefully will have one by tomorrow.

**Comment by sydney-runkle** [comment]
I've opened a PR with a fix, should be included in v2.9 by the end of the month!
